File: М11_ClusterVMS/clustervms-go/cluster/objectstore.go
package cluster

// The cluster's object store — large, rare, never queried: the restore
// point (and, since the heartbeat left raft, the heartbeat).
//
// Three adapters with one contract. HttpObjectStore PUTs and GETs against
// any endpoint with plain HTTP object semantics (MinIO with a bucket
// policy, nginx with dav). FsObjectStore is a directory — the tests, and a
// bench with a shared mount. S3ObjectStore (s3.go) signs.

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ObjectStore: Get returns (nil, nil) for an object that does not exist.
type ObjectStore interface {
	Put(key string, data []byte) error
	Get(key string) ([]byte, error)
}

type HttpObjectStore struct {
	Base   string
	Client *http.Client
}

func NewHttpObjectStore(base string) *HttpObjectStore {
	return &HttpObjectStore{Base: strings.TrimRight(base, "/"), Client: &http.Client{Timeout: 10 * time.Second}}
}

func (h *HttpObjectStore) Put(key string, data []byte) error {
	req, err := http.NewRequest("PUT", h.Base+"/"+key, strings.NewReader(string(data)))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/octet-stream")
	resp, err := h.Client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 && resp.StatusCode != 201 && resp.StatusCode != 204 {
		return fmt.Errorf("PUT %s: HTTP %d", key, resp.StatusCode)
	}
	return nil
}

func (h *HttpObjectStore) Get(key string) ([]byte, error) {
	resp, err := h.Client.Get(h.Base + "/" + key)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode == 404 {
		return nil, nil
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("GET %s: HTTP %d", key, resp.StatusCode)
	}
	return io.ReadAll(resp.Body)
}

type FsObjectStore struct{ Root string }

func NewFsObjectStore(root string) (*FsObjectStore, error) {
	return &FsObjectStore{root}, os.MkdirAll(root, 0o755)
}

func (f *FsObjectStore) Put(key string, data []byte) error {
	p := filepath.Join(f.Root, key)
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(p+".tmp", data, 0o644); err != nil {
		return err
	}
	return os.Rename(p+".tmp", p) // an object appears whole or not at all
}

func (f *FsObjectStore) Get(key string) ([]byte, error) {
	b, err := os.ReadFile(filepath.Join(f.Root, key))
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	return b, err
}

// OpenStore: file:///path · http(s)://host/bucket (anonymous) ·
// s3+http(s)://host/bucket?region=r (SigV4, credentials from
// AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY — on a Node, from its Variable).
func OpenStore(raw string) (ObjectStore, error) {
	switch {
	case strings.HasPrefix(raw, "s3+http://"), strings.HasPrefix(raw, "s3+https://"):
		u, err := url.Parse(raw[3:])
		if err != nil {
			return nil, err
		}
		region := u.Query().Get("region")
		if region == "" {
			region = "us-east-1"
		}
		return NewS3ObjectStore(u.Scheme+"://"+u.Host, strings.Trim(u.Path, "/"), region,
			os.Getenv("AWS_ACCESS_KEY_ID"), os.Getenv("AWS_SECRET_ACCESS_KEY"))
	case strings.HasPrefix(raw, "http://"), strings.HasPrefix(raw, "https://"):
		return NewHttpObjectStore(raw), nil
	case strings.HasPrefix(raw, "file://"):
		return NewFsObjectStore(strings.TrimPrefix(raw, "file://"))
	}
	return NewFsObjectStore(raw)
}

File: М11_ClusterVMS/clustervms/cluster/objectstore.py
"""The cluster's object store — М10's `vmsplatform.objects.ObjectStore`
contract on MinIO or S3: heartbeats (frequent, never queried by key) and the
snapshots the domain's read model is built from.

Two adapters with one contract. `HttpObjectStore` PUTs and GETs against any
endpoint that accepts plain HTTP object semantics (MinIO with a bucket
policy, nginx with dav, an S3 presigned pattern behind a proxy).
`FsObjectStore` is a directory — the tests, and a bench with a shared mount.
An S3 adapter with signed requests is a twenty-line boto3 wrapper on the
same two methods; it is not here because the Node image carries no boto3.
"""
from __future__ import annotations

import os
import urllib.error
import urllib.request
from typing import Protocol


class ObjectStore(Protocol):
    def put(self, key: str, data: bytes) -> None: ...
    def get(self, key: str) -> bytes | None: ...
    def list(self, prefix: str) -> list[str]: ...


class HttpObjectStore:
    def __init__(self, base_url: str, timeout: float = 10.0):
        self.base = base_url.rstrip("/")
        self.timeout = timeout

    def put(self, key: str, data: bytes) -> None:
        req = urllib.request.Request(f"{self.base}/{key}", data=data, method="PUT")
        req.add_header("Content-Type", "application/octet-stream")
        with urllib.request.urlopen(req, timeout=self.timeout) as r:
            if r.status not in (200, 201, 204):
                raise IOError(f"PUT {key}: {r.status}")

    def get(self, key: str) -> bytes | None:
        try:
            with urllib.request.urlopen(f"{self.base}/{key}", timeout=self.timeout) as r:
                return r.read()
        except urllib.error.HTTPError as e:
            if e.code == 404:
                return None
            raise

    def list(self, prefix: str) -> list[str]:
        raise NotImplementedError("plain HTTP has no listing; use s3+http:// for the heartbeat prefix")


class FsObjectStore:
    def __init__(self, root: str):
        self.root = root
        os.makedirs(root, exist_ok=True)

    def put(self, key: str, data: bytes) -> None:
        p = os.path.join(self.root, key)
        os.makedirs(os.path.dirname(p), exist_ok=True)
        with open(p + ".tmp", "wb") as f:
            f.write(data)
        os.replace(p + ".tmp", p)                  # an object appears whole or not at all

    def get(self, key: str) -> bytes | None:
        p = os.path.join(self.root, key)
        if not os.path.exists(p):
            return None
        with open(p, "rb") as f:
            return f.read()

    def list(self, prefix: str) -> list[str]:
        out = []
        for d, _, files in os.walk(self.root):
            for f in files:
                if f.endswith(".tmp"):
                    continue
                key = os.path.relpath(os.path.join(d, f), self.root)
                if key.startswith(prefix):
                    out.append(key)
        return sorted(out)


def open_store(url: str) -> ObjectStore:
    """file:///path · http(s)://host/bucket (anonymous) · s3+http(s)://host/bucket?region=r (SigV4,
    credentials from AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY — on a Node, from its Variable)."""
    if url.startswith(("s3+http://", "s3+https://")):
        from urllib.parse import parse_qs, urlsplit
        from .s3 import S3ObjectStore
        u = urlsplit(url[3:])
        region = parse_qs(u.query).get("region", ["us-east-1"])[0]
        return S3ObjectStore(f"{u.scheme}://{u.netloc}", u.path.strip("/"), region)
    if url.startswith(("http://", "https://")):
        return HttpObjectStore(url)
    if url.startswith("file://"):
        return FsObjectStore(url[len("file://"):])
    return FsObjectStore(url)
